url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/Data.java

```diff
@@ -0,0 +1,89 @@
+package utils;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Hashtable;
+
+import protocols.specifications.DBProtocol;
+
+import attributes.Attribute;
+import attributes.DistanceFraudProbability;
+import attributes.MafiaFraudProbability;
+import attributes.TerroristFraudProbability;
+import attributes.relations.ProbabilityRelation;
+import attributes.scales.LogScale;
+
+public class Data {
+	
+	/*Trujillo- Jun 3, 2014
+	 * Create a data file for each protocol*/
+	public static void printData(DBProtocol[] protocols, Attribute[] attributes, int maxN, int increment) throws IOException{
+		String suffix = "";
+		String newLine = System.getProperty("line.separator");
+		for (Attribute attribute : attributes) {
+			suffix += "-"+attribute.getName();
+		}
+		for (DBProtocol p : protocols) {
+			FileWriter writer = new FileWriter(new File(p.getIdentifier()+suffix+"-increment-"+increment+".DAT"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced at lines 29, 31, 33, 35, 37, 39. The resource is closed at line 41. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).